gobblin-temporal/src/main/java/org/apache/gobblin/temporal/yarn/YarnService.java

```diff
@@ -411,10 +330,24 @@ protected synchronized void requestContainersForWorkerProfile(WorkerProfile work
     requestContainers(numContainers, Resource.newInstance(containerMemoryMbs, containerCores), Optional.of(allocationRequestId));
   }
 
-  private void requestContainer(Optional<String> preferredNode, Optional<Resource> resourceOptional) {
-    Resource desiredResource = resourceOptional.or(Resource.newInstance(
-        this.requestedContainerMemoryMbs, this.requestedContainerCores));
-    requestContainer(preferredNode, desiredResource, Optional.absent());
+  protected synchronized void releaseContainersForWorkerProfile(String profileName, int numContainers) {
+    Iterator<Map.Entry<ContainerId, ContainerInfo>> containerMapIterator = this.containerMap.entrySet().iterator();
+    while (containerMapIterator.hasNext() && numContainers > 0) {
+      Map.Entry<ContainerId, ContainerInfo> entry = containerMapIterator.next();
+      if (entry.getValue().getWorkerProfile().getName().equals(profileName)) {
+        ContainerId containerId = entry.getKey();
+        LOGGER.info("Releasing container {} running profile {}", containerId, WorkforceProfiles.renderName(profileName));
+        // Record that this container was explicitly released so that a new one is not spawned to replace it
+        // Put the container id in the releasedContainerCache before releasing it so that handleContainerCompletion()
+        // can check for the container id and skip spawning a replacement container.
+        // Note that this is the best effort since these are asynchronous operations and a container may abort concurrently
+        // with the release call. So in some cases a replacement container may have already been spawned before
+        // the container is put into the black list.
+        this.releasedContainerCache.put(containerId, "");
+        this.amrmClientAsync.releaseAssignedContainer(containerId);
+        numContainers--;
+      }
+    }
```

Comment:
please add a log here for how many containers were intended to be released and how many were actually released from this method